File: examples/mobile/android/AionMCPRepository.kt
package com.example.aionmcp.data.repository

import com.example.aionmcp.data.api.AionMCPApiService
import com.example.aionmcp.data.models.*
import kotlinx.coroutines.delay
import java.io.IOException

/**
 * Repository for AionMCP API operations
 * Implements error handling, retry logic, and data caching
 */
class AionMCPRepository(
    private val api: AionMCPApiService
) {
    // Simple in-memory cache
    private var toolsCache: List<Tool>? = null
    private var toolsCacheTimestamp: Long = 0
    private val cacheValidityMs = 5 * 60 * 1000L // 5 minutes
    
    /**
     * Check server health
     */
    suspend fun checkHealth(): Result<HealthResponse> = safeApiCall {
        api.getHealth()
    }
    
    /**
     * List all available tools with caching
     */
    suspend fun listTools(forceRefresh: Boolean = false): Result<List<Tool>> {
        // Return cached data if valid
        if (!forceRefresh && isCacheValid()) {
            toolsCache?.let { return Result.success(it) }
        }
        
        return safeApiCall {
            val response = api.listTools()
            toolsCache = response.tools
            toolsCacheTimestamp = System.currentTimeMillis()
            response.tools
        }
    }
    
    /**
     * Invoke a tool with retry logic
     */
    suspend fun invokeTool(
        toolName: String,
        args: Map<String, Any>,
        maxRetries: Int = 3
    ): Result<ToolInvokeResponse> {
        return retryApiCall(maxRetries) {
            api.invokeTool(ToolInvokeRequest(toolName, args))
        }
    }
    
    /**
     * Get learning statistics
     */
    suspend fun getLearningStats(): Result<LearningStatsResponse> = safeApiCall {
        api.getLearningStats()
    }
    
    /**
     * Get learning insights
     */
    suspend fun getInsights(
        type: String? = null,
        priority: String? = null
    ): Result<List<Insight>> = safeApiCall {
        api.getInsights(type, priority).insights
    }
    
    /**
     * Get insights for a specific tool
     */
    suspend fun getToolInsights(toolName: String): Result<List<Insight>> = safeApiCall {
        api.getToolInsights(toolName).insights
    }
    
    /**
     * Trigger pattern analysis
     */
    suspend fun analyzePatterns(): Result<Map<String, Any>> = safeApiCall {
        api.analyzePatterns()
    }
    
    /**
     * Clear the tools cache
     */
    fun clearCache() {
        toolsCache = null
        toolsCacheTimestamp = 0
    }
    
    /**
     * Check if cached data is still valid
     */
    private fun isCacheValid(): Boolean {
        return toolsCache != null && 
               (System.currentTimeMillis() - toolsCacheTimestamp) < cacheValidityMs
    }
    
    /**
     * Safe API call wrapper with error handling
     */
    private suspend fun <T> safeApiCall(
        apiCall: suspend () -> T
    ): Result<T> {
        return try {
            Result.success(apiCall())
        } catch (e: IOException) {
            Result.failure(NetworkException("Network error: ${e.message}", e))
        } catch (e: Exception) {
            Result.failure(ApiException("API error: ${e.message}", e))
        }
    }
    
    /**
     * API call with retry logic
     */
    private suspend fun <T> retryApiCall(
        maxRetries: Int,
        apiCall: suspend () -> T
    ): Result<T> {
        var lastException: Exception? = null
        
        repeat(maxRetries) { attempt ->
            try {
                return Result.success(apiCall())
            } catch (e: IOException) {
                lastException = NetworkException("Network error: ${e.message}", e)
                // Exponential backoff
                if (attempt < maxRetries - 1) {
                    delay(1000L * (attempt + 1))
                }
            } catch (e: Exception) {
                // Don't retry on non-network errors
                return Result.failure(ApiException("API error: ${e.message}", e))
            }
        }
        
        return Result.failure(lastException ?: Exception("Unknown error"))
    }
}

/**
 * Custom exception for network errors
 */
class NetworkException(message: String, cause: Throwable? = null) : Exception(message, cause)

/**
 * Custom exception for API errors
 */
class ApiException(message: String, cause: Throwable? = null) : Exception(message, cause)

File: examples/mobile/ios/AionMCPRepository.swift
import Foundation

/// Repository for AionMCP operations with error handling and caching
actor AionMCPRepository {
    private let client: AionMCPClient
    private var toolsCache: [Tool]?
    private var cacheTimestamp: Date?
    private let cacheValiditySeconds: TimeInterval = 300 // 5 minutes
    
    init(client: AionMCPClient) {
        self.client = client
    }
    
    // MARK: - Health
    
    func checkHealth() async throws -> HealthResponse {
        return try await client.getHealth()
    }
    
    // MARK: - Tools
    
    func listTools(forceRefresh: Bool = false) async throws -> [Tool] {
        // Return cached data if valid
        if !forceRefresh, let cached = toolsCache, isCacheValid() {
            return cached
        }
        
        // Fetch from API
        let response = try await client.listTools()
        toolsCache = response.tools
        cacheTimestamp = Date()
        return response.tools
    }
    
    func invokeTool(name: String, args: [String: Any], maxRetries: Int = 3) async throws -> ToolInvokeResponse {
        return try await retryOperation(maxRetries: maxRetries) {
            try await self.client.invokeTool(name: name, args: args)
        }
    }
    
    // MARK: - Learning
    
    func getLearningStats() async throws -> LearningStatsResponse {
        return try await client.getLearningStats()
    }
    
    func getInsights(type: String? = nil, priority: String? = nil) async throws -> [Insight] {
        let response = try await client.getInsights(type: type, priority: priority)
        return response.insights
    }
    
    func getToolInsights(toolName: String) async throws -> [Insight] {
        let response = try await client.getToolInsights(toolName: toolName)
        return response.insights
    }
    
    func analyzePatterns() async throws -> [String: Any] {
        return try await client.analyzePatterns()
    }
    
    // MARK: - Cache Management
    
    func clearCache() {
        toolsCache = nil
        cacheTimestamp = nil
    }
    
    private func isCacheValid() -> Bool {
        guard let timestamp = cacheTimestamp else { return false }
        return Date().timeIntervalSince(timestamp) < cacheValiditySeconds
    }
    
    // MARK: - Retry Logic
    
    private func retryOperation<T>(
        maxRetries: Int,
        operation: @escaping () async throws -> T
    ) async throws -> T {
        var lastError: Error?
        
        for attempt in 0..<maxRetries {
            do {
                return try await operation()
            } catch let error as AionMCPError {
                lastError = error
                
                // Don't retry on certain errors
                switch error {
                case .unauthorized:
                    throw error
                case .networkError:
                    // Only retry network errors
                    if attempt < maxRetries - 1 {
                        // Exponential backoff
                        let delay = UInt64(1_000_000_000 * (attempt + 1))
                        try await Task.sleep(nanoseconds: delay)
                    }
                default:
                    throw error
                }
            } catch {
                lastError = error
                if attempt < maxRetries - 1 {
                    let delay = UInt64(1_000_000_000 * (attempt + 1))
                    try await Task.sleep(nanoseconds: delay)
                }
            }
        }
        
        throw lastError ?? AionMCPError.invalidResponse
    }
}
